{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 8.文件处理\n",
    "\n",
    "文件是保存在计算机存储设备上的一些信息或数据。比如你的音乐文件，视频文件，文本文件。Python提供了一些简单的方式操纵文件。通常我们把文件分为两类，文本文件和二进制文件。文本文件是简单的文本，二进制文件包含了只有计算机可读的二进制数据。\n",
    "\n",
    "## 8.1.文件操作\n",
    "\n",
    "### 文件打开\n",
    "\n",
    "我们使用 open() 函数打开文件。它需要两个参数，第一个参数是文件路径或文件名，第二个是文件的打开模式。模式通常是下面这样的：\n",
    "\n",
    "    \"r\"，以只读模式打开，你只能读取文件但不能编辑/删除文件的任何内容\n",
    "    \"w\"，以写入模式打开，如果文件存在将会删除里面的所有内容，然后打开这个文件进行写入\n",
    "    \"a\"，以追加模式打开，写入到文件中的任何数据将自动添加到末尾\n",
    "\n",
    "默认的模式为只读模式，也就是说如果你不提供任何模式，open() 函数将会以只读模式打开文件。\n",
    "```python\n",
    "file = open(\"sample.txt\")\n",
    "```\n",
    "\n",
    "### 文件关闭\n",
    "\n",
    "打开文件后我们应该总是关闭文件。我们使用方法 close() 完成这个操作。\n",
    "```python\n",
    "file.close()\n",
    "```\n",
    "要始终确保你显式关闭每个打开的文件，一旦它的工作完成你没有任何理由保持打开文件。因为程序能打开的文件数量是有上限的。如果你超出了这个限制，没有任何可靠的方法恢复，因此程序可能会崩溃。每个打开的文件关联的数据结构（文件描述符/句柄/文件锁...）都要消耗一些主存资源。因此如果许多打开的文件没用了你可以结束大量的内存浪费，并且文件打开时始终存在数据损坏或丢失的可能性。\n",
    "\n",
    "\n",
    "### 文件读取\n",
    "使用 read() 方法一次性读取整个文件。\n",
    "```python\n",
    "fobj = open(\"sample.txt\")\n",
    "fobj.read()\n",
    "```\n",
    "如果你再一次调用 read()，它会返回空字符串因为它已经读取完整个文件。\n",
    "\n",
    "read(size) 有一个可选的参数 size，用于指定字符串长度。如果没有指定 size 或者指定为负数，就会读取并返回整个文件。当文件大小为当前机器内存两倍时，就会产生问题。反之，会尽可能按比较大的 size 读取和返回数据。\n",
    "\n",
    "readline() 能帮助你每次读取文件的一行。 readlines() 方法读取所有行到一个列表中。\n",
    "```python\n",
    "fobj.readline()\n",
    "fobj.readlines()\n",
    "```\n",
    "\n",
    "### 文件写入\n",
    "让我们通过 write() 方法打开一个文件然后我们随便写入一些文本。\n",
    "```python\n",
    "fobj = open(\"ircnicks.txt\", 'w')\n",
    "fobj.write('powerpork\\n')\n",
    "fobj.write('indrag\\n')\n",
    "fobj.write('mishti\\n')\n",
    "fobj.write('sankarshan')\n",
    "fobj.close()\n",
    "```\n",
    "\n",
    "## 8.2.使用with语句\n",
    "\n",
    "在实际情况中，我们应该尝试使用 with 语句处理文件对象，它会在文件用完后会自动关闭，就算发生异常也没关系。它是 try-finally 块的简写：\n",
    "```python\n",
    "with open('sample.txt') as fobj:\n",
    "    for line in fobj:\n",
    "        print(line, end = '')\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9.异常处理\n",
    "\n",
    "在程序执行过程中发生的任何错误都是异常。每个异常显示一些相关的错误信息，比如你在 Python3 中使用 Python2 独有的语法就会发生 SyntaxError。\n",
    "\n",
    "当有人试图访问一个未定义的变量则会发生 NameError。\n",
    "TypeError 也是一种经常出现的异常。当操作或函数应用于不适当类型的对象时引发，一个常见的例子是对整数和字符串做加法。\n",
    "```python\n",
    "print(1 + \"kushal\")\n",
    "```\n",
    "\n",
    "## 异常处理\n",
    "我们使用 try...except 块来处理任意异常。基本的语法像这样：\n",
    "```python\n",
    "try:\n",
    "    statements to be inside try clause\n",
    "    statement2\n",
    "    statement3\n",
    "except ExceptionName:\n",
    "    statements to evaluated in case of ExceptionName happens\n",
    "```\n",
    "它以如下方式工作：\n",
    "\n",
    "* 首先，执行 try 子句 （在 try 和 except 关键字之间的部分）。\n",
    "* 如果没有异常发生，except 子句 在 try 语句执行完毕后就被忽略了。\n",
    "* 如果在 try 子句执行过程中发生了异常，那么该子句其余的部分就会被忽略。\n",
    "\n",
    "## 抛出异常\n",
    "使用 raise 语句抛出一个异常。我们可以捕获任何其它普通异常一样，来捕获抛出的这些异常。\n",
    "```python\n",
    "raise ValueError(\"A value error happened.\")\n",
    "```\n",
    "\n",
    "## 定义最终处理操作\n",
    "\n",
    "try 语句还有另一个可选的 finally 子句，目的在于定义在任何情况下都一定要执行的功能。\n",
    "```python\n",
    "try:\n",
    "    raise KeyboardInterrupt\n",
    "finally:\n",
    "    print('Goodbye, world!')\n",
    "```\n",
    "不管有没有发生异常，finally 子句 在程序离开 try 后都一定会被执行。当 try 语句中发生了未被 except 捕获的异常（或者它发生在 except 或 else 子句中），在 finally 子句执行完后它会被重新抛出。\n",
    "\n",
    "在真实场景的应用程序中，finally 子句用于释放外部资源（文件或网络连接之类的），无论它们的使用过程中是否出错。"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
